### PR TITLE
feature(views): allow providing alternative views for list items

### PR DESCRIPTION
**Pull request by hypeJunction**
Alternative view to render list items can now be passed as 'item_view' parameter to list generating views and functions

Refs #7899

### CODE REVIEW

File: engine/lib/views.php [hunk 1]
```diff
@@ -813,25 +813,25 @@ function elgg_view_entity(\ElggEntity $entity, $vars = array(), $bypass = false,
...
-	if (empty($subtype)) {
-		$subtype = 'default';
+	$entity_subtype = $entity->getSubtype();
+	if (empty($entity_subtype)) {
+		$entity_subtype = 'default';
 	}
 
+	$entity_views = array(
+		$entity->view,
+		$entity->getVolatileData('view'),
```
Comment (mrclay) [inline engine/lib/views.php]: Put this first!

File: engine/classes/Elgg/ViewsService.php [hunk 2]
```diff
@@ -405,7 +405,10 @@ private function renderViewFile($view, array $vars, $viewtype, $issue_missing_no
 	 */
 	public function viewExists($view, $viewtype = '', $recurse = true) {
 		
-
+		if (!is_string($view)) {
```
Comment (mrclay) [inline engine/classes/Elgg/ViewsService.php]: If empty($view)

File: engine/lib/views.php (hunk 1 above)
Comment (mrclay) [inline engine/lib/views.php]: 2nd in list should be $vars['entity_view'] in case it's passed in.

Comment (mrclay) [inline engine/lib/views.php]: ...then we can make sure the entity_view option is passed down from elgg_list_entities.

Comment (hypeJunction) [inline engine/lib/views.php]: Yep, doing that. I will keep volatileData, and add $vars['entity_view']

Comment (hypeJunction) [inline engine/lib/views.php]: So, lists can have entity_view, and then if some item needs to be rendered differently, the plugin can set volatileData on that entity

Comment (hypeJunction) [inline engine/lib/views.php]: Or should it be 'item_view' to be in line with 'items'?

Comment (mrclay) [inline engine/lib/views.php]: `item_view` sounds good. Could be used viewing annotations/other things. I could even see calling elgg_list_entities() with `'callback' => ''` to explicitly avoid the overhead of entity instantiation, but that's probably not something we want to go out of our way to support.

File: engine/lib/views.php [hunk 3]
```diff
@@ -813,25 +813,25 @@ function elgg_view_entity(\ElggEntity $entity, $vars = array(), $bypass = false,
...
+		"$entity_type/default",
+	);
+
 	$contents = '';
-	if (elgg_view_exists("$entity_type/$subtype")) {
-		$contents = elgg_view("$entity_type/$subtype", $vars, $bypass, $debug);
-	} else {
-		$contents = elgg_view("$entity_type/default", $vars, $bypass, $debug);
+	foreach ($entity_views as $view) {
+		if (elgg_view_exists($view)) {
```
Comment (mrclay) [inline engine/lib/views.php]: Might as well use `$this->viewExists`

Comment (mrclay) [inline engine/lib/views.php]: oh, sorry, forgot this isn't the ViewService

File: engine/lib/views.php (hunk 1 above)
Comment (hypeJunction) [inline engine/lib/views.php]: I went for explicit 'entity_view', 'annotation_view', 'river_view' to avoid passing wrong object type in a view.

File: engine/lib/views.php [hunk 4]
```diff
@@ -1342,20 +1373,19 @@ function elgg_view_tagcloud(array $options = array()) {
  *
  * @param \ElggEntity|\ElggAnnotation $item
  * @param array  $vars Additional parameters for the rendering
- *
+ *      'entity_view'     Alternative view to render ElggEntity
+ *      'annotation_view' Alternative view to render ElggAnnotation
+ *      'river_view'      Alternative view to render ElggRiverItem
```
Comment (mrclay) [inline engine/lib/views.php]: `river_item_view`?

File: engine/lib/views.php [hunk 5]
```diff
@@ -1342,20 +1373,19 @@ function elgg_view_tagcloud(array $options = array()) {
...
  * @return string
  * @since 1.8.0
  * @access private
  */
 function elgg_view_list_item($item, array $vars = array()) {
-	global $CONFIG;
-
-	$type = $item->getType();
-	if (in_array($type, $CONFIG->entity_types)) {
```
Comment (mrclay) [inline engine/lib/views.php]: I know I suggested refactoring an elgg_view_exists() call before, I was wrong. Putting unrelated refactoring into this PR will make it harder to evaluate. This is good but could go in another PR.

File: engine/lib/views.php (hunk 4 above)
Comment (mrclay) [inline engine/lib/views.php]: For description, how about:

* "View used to render entities (optional)"
* "View used to render annotations (optional)"
* "View used to render river items (optional)"

**Comment by mrclay** [comment]
This is starting to remind me of #7360, but simpler. See also #7376

**Comment by hypeJunction** [comment]
I have moved all the unrelated stuff into separate PRs

**Comment by mrclay** [comment]
Much clearer, thanks

**Comment by hypeJunction** [comment]
@mrclay Now using a single 'item_view' parameter

**Comment by mrclay** [comment]
@ewinslow I know we've talked about viewtype/format overhauls that would help in this area but I think this is a simple and easy-to-understand solution to this use case. Thoughts?

If we agree to move forward, this should get some RST docs.

**Comment by jdalsem** [comment]
LGTM; with the docs this should be ready to go

**Comment by Srokap** [comment]
We also have this old view metadata support: https://github.com/Elgg/Elgg/blob/1.10/engine/lib/views.php#L815

I really wonder if anyone actually uses that. I wouldn't really mind dropping support for it.

**Comment by jdalsem** [comment]
Never used it personally, but dropping it should be another ticket :)

**Comment by hypeJunction** [comment]
RST docs are in. Anything anyone would like to add?

File: docs/guides/views.rst [hunk 6]
```diff
@@ -295,6 +295,34 @@ If your entity list will display the entity owners, you can improve performance
 
 See also :doc:`check this page out first </design/database>`.
 
+Since 1.11, you can define an alternative view to render list items using ```'item_view'``` parameter.
```
Comment (juho-jaakkola) [inline docs/guides/views.rst]: Do we need the version number here? After all this will not get added to 1.10 docs.

Comment (juho-jaakkola) [inline docs/guides/views.rst]: IMO we should avoid having hard-coded version numbers in the docs.

Comment (ewinslow) [inline docs/guides/views.rst]: I think "As of {version}" and "As of {date}" type statements are OK, because that information will never go out of date, but might still be useful to someone reading the latest documentation (e.g. if their version of Elgg is not updated to the latest docs).

Comment (hypeJunction) [inline docs/guides/views.rst]: Is a there a consensus on this? I don't want to this 'As of string' holding back the entire PR

Comment (jdalsem) [inline docs/guides/views.rst]: i'm fine with leaving 1.11 in, as 1.10 devs also read the latest docs probably, so it is good to know when to use this new feature

Comment (juho-jaakkola) [inline docs/guides/views.rst]: Not a blocker.

Comment (hypeJunction) [inline docs/guides/views.rst]: Ok. Someone press the merge button then ;)

**Comment by hypeJunction** [comment]
Yay. I am looking forward to this. I will make PRs to rewrite some of the
boilerplate code floating around
